### PR TITLE
Inf3

### PR DESCRIPTION
**Pull request by wilkeraziz**

### CODE REVIEW

**Comment by wilkeraziz** [comment]
This commit changes the general InferenceModel and InferenceEncoder API slightly, so I consider this part of the master. It also adds a special type of inference model (which is not essential in the master, but can't hurt).

File: demo/hparams/aevnmt_supervised_flickr.json [hunk 1]
```diff
@@ -29,8 +29,8 @@
     "model_type": "aevnmt",
     "num_dec_layers": 1,
     "num_enc_layers": 1,
-    "num_epochs": 10,
-    "patience": 2,
+    "num_epochs": 15,
+    "patience": 5,
```
Comment (Roxot) [inline demo/hparams/aevnmt_supervised_flickr.json]: Why did you change these? This is not what the downloaded demo models will be trained with, so it's better to leave them as they were.

File: demo/aevnmt-translate.sh [hunk 2]
```diff
@@ -6,7 +6,7 @@ TAG=flickr
 SRC=de
 TGT=en
 DATA=${TAG}/data
-TRAINED_MODEL=${TAG}/models/aevnmt/supervised
+TRAINED_MODEL=${TAG}/models/aevnmt/supervised.${SRC}-${TGT}
```
Comment (Roxot) [inline demo/aevnmt-translate.sh:9]: Will this not break existing downloaded models?

File: demo/aevnmt-training.sh [hunk 3]
```diff
@@ -4,7 +4,7 @@ TAG=flickr
 SRC=de
 TGT=en
 DATA=${TAG}/data
-OUTPUT=${TAG}/models/aevnmt/supervised
+OUTPUT=${TAG}/models/aevnmt/supervised.${SRC}-${TGT}
```
Comment (Roxot) [inline demo/aevnmt-training.sh:7]: Will this not break existing downloaded models?

**Review by Roxot — CHANGES_REQUESTED** [review]

File: demo/aevnmt-translate.sh (hunk 2 above)
Comment (wilkeraziz) [reply demo/aevnmt-translate.sh:9]: Compatibility with existing models is long gone (though architecture blocks may remain the same, refactoring affected the parameters' names). The best thing would be to retrain those models (perhaps have different trained models for different versions of the code.

File: demo/hparams/aevnmt_supervised_flickr.json (hunk 1 above)
Comment (wilkeraziz) [reply demo/hparams/aevnmt_supervised_flickr.json]: Accidental change.

File: demo/aevnmt-translate.sh (hunk 2 above)
Comment (Roxot) [reply demo/aevnmt-translate.sh:9]: I see, I'll have a look at that later then.

File: demo/hparams/aevnmt_supervised_flickr.json (hunk 1 above)
Comment (Roxot) [reply demo/hparams/aevnmt_supervised_flickr.json]: After you revert those feel free to merge.
